core/primitives-core/src/account.rs

```diff
@@ -256,14 +488,134 @@ mod tests {
         };
         let mut old_bytes = &borsh::to_vec(&old_account).unwrap()[..];
         let new_account = <Account as BorshDeserialize>::deserialize(&mut old_bytes).unwrap();
-        assert_eq!(new_account.amount, old_account.amount);
-        assert_eq!(new_account.locked, old_account.locked);
-        assert_eq!(new_account.code_hash, old_account.code_hash);
-        assert_eq!(new_account.storage_usage, old_account.storage_usage);
+
+        assert_eq!(new_account.amount(), old_account.amount);
+        assert_eq!(new_account.locked(), old_account.locked);
+        assert_eq!(new_account.code_hash(), old_account.code_hash);
+        assert_eq!(new_account.storage_usage(), old_account.storage_usage);
         assert_eq!(new_account.version, AccountVersion::V1);
+
         let mut new_bytes = &borsh::to_vec(&new_account).unwrap()[..];
         let deserialized_account =
             <Account as BorshDeserialize>::deserialize(&mut new_bytes).unwrap();
         assert_eq!(deserialized_account, new_account);
     }
+
+    #[test]
+    fn test_account_v1_serde_serialization() {
+        let account = Account {
+            amount: 10_000_000,
+            locked: 100_000,
+            #[cfg(feature = "protocol_feature_nonrefundable_transfer_nep491")]
+            nonrefundable: 0,
+            code_hash: CryptoHash::default(),
+            storage_usage: 1000,
+            version: AccountVersion::V1,
+        };
+        let serialized_account = serde_json::to_string(&account).unwrap();
+        let deserialized_account: Account = serde_json::from_str(&serialized_account).unwrap();
+        assert_eq!(deserialized_account, account);
+    }
+
+    #[cfg(feature = "protocol_feature_nonrefundable_transfer_nep491")]
+    /// Serialization of account V1 with non-refundable amount greater than 0 would pass without an error,
+    /// but an error would be raised on deserialization of such invalid data.
```

Comment:
We cannot really create new V1 accounts. It is possible for tests in this module, but for client these fields are private and it has to use`Account::new()`. `Account::new` always sets version to V2, and if nonrefundable feature is disabled it asserts that nonrefundable amount is 0.
So this test documents somewhat impossible situation if we already had an account V1 with nonrefundable amount greater than 0.

Reply:
I've just updated `Account::new()` so that it explicitly checks that nonrefundable is 0 if the default account version is V1.

Reply:
> If nonrefundable feature is enabled, `Account::new()` sets version to V2, otherwise it asserts that nonrefundable amount is 0.

Hm, I have to admit, I probably didn't think about this case properly when I wrote the reference implementation. But I believe to stabilize this feature, we must make account creation version dependent on the protocol version. Otherwise, validator nodes with the new binary will immediately start producing accounts of version V2, even if the protocol upgrade voting hasn't completed yet. That would cause different state roots for different nearcore binaries, and therefore validators would miss their chunks/blocks and get kicked out in the next epoch.

Does that make sense?